File: solutions/cpp/next_permutation.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::swap;
using std::uniform_int_distribution;
using std::vector;

// @include
vector<int> NextPermutation(vector<int> perm) {
  // Find the first entry from the right that is smaller than the entry
  // immediately after it.
  auto inversion_point = is_sorted_until(perm.rbegin(), perm.rend());
  if (inversion_point == perm.rend()) {
    // perm is sorted in decreasing order, so it's the last permutation.
    return {};
  }

  // Swap the entry referenced by inversion_point with smallest entry
  // appearing after inversion_point that is greater than the entry referenced
  // by inversion_point:
  //
  // 1.) Find the smallest entry after inversion_point that's greater than the
  //     entry referenced by inversion_point. Since perm must be sorted in
  //     decreasing order after inversion_point, we can use a fast algorithm
  //     to find this entry.
  auto least_upper_bound =
      upper_bound(perm.rbegin(), inversion_point, *inversion_point);

  // 2.) Perform the swap.
  iter_swap(inversion_point, least_upper_bound);

  // Reverse the subarray that follows inversion_point.
  reverse(perm.rbegin(), inversion_point);
  return perm;
}
// @exclude

int main(int argc, char *argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    vector<int> perm;
    if (argc > 2) {
      for (size_t i = 1; i < argc; ++i) {
        perm.emplace_back(atoi(argv[i]));
      }
    } else {
      uniform_int_distribution<int> dis(1, 100);
      int n = (argc == 2 ? atoi(argv[1]) : dis(gen));
      uniform_int_distribution<int> n_dis(0, n - 1);
      generate_n(back_inserter(perm), n, [&] { return n_dis(gen); });
    }

    vector<int> ans(NextPermutation(perm));
    // Use built-in function verification.
    bool has_next_one = next_permutation(perm.begin(), perm.end());
    assert((ans.size() == 0 && !has_next_one) ||
           equal(ans.cbegin(), ans.cend(), perm.cbegin(), perm.cend()));
  }
  return 0;
}

File: solutions/python/next_permutation.py
import itertools
import random
import sys


# @include
def next_permutation(perm):
    # Find the first entry from the right that is smaller than the entry
    # immediately after it.
    inversion_point = len(perm) - 2
    while (inversion_point >= 0 and
           perm[inversion_point] >= perm[inversion_point + 1]):
        inversion_point -= 1
    if inversion_point == -1:
        return []  # perm is the last permutation.

    # Swap the smallest entry after index inversion_point that is greater than
    # perm[inversion_point]. Since entries in perm are decreasing after
    # inversion_point, if we search in reverse order, the first entry that is
    # greater than perm[inversion_point] is the entry to swap with.
    for i in reversed(range(inversion_point + 1, len(perm))):
        if perm[i] > perm[inversion_point]:
            perm[inversion_point], perm[i] = perm[i], perm[inversion_point]
            break

    # Entries in perm must appear in decreasing order after inversion_point,
    # so we simply reverse these entries to get the smallest dictionary order.
    perm[inversion_point + 1:] = reversed(perm[inversion_point + 1:])
    return perm
# @exclude


def main():
    for _ in range(1000):
        if len(sys.argv) > 2:
            perm = list(map(int, sys.argv[1:]))
        else:
            n = int(
                sys.argv[1]) if len(
                sys.argv) == 2 else random.randint(
                1,
                10)
            perm = [random.randrange(n) for _ in range(n)]

        gold = list(itertools.permutations(sorted(perm)))
        for i, p in reversed(list(enumerate(gold))):
            if list(p) == perm:
                target = i
                break
        ans = next_permutation(perm)
        assert (target == len(gold) - 1 and ans == []
                ) or (list(gold[target + 1]) == ans)


if __name__ == '__main__':
    main()
